config/src/test/scala/com/typesafe/config/impl/ConfigSubstitutionTest.scala

```diff
@@ -924,7 +924,7 @@ class ConfigSubstitutionTest extends TestUtils {
 
     @Test
     def substSelfReferenceIndirect() {
-        val obj = parseObject("""a=1, b=${a}, a=${b}""")
+        val obj = parseObject("""b=${a}, a=${b}""")
```

Comment:
Java 7 parses as:

    {b=ConfigReference(${a}), a=ConfigDelayedMerge(1,${b})}

Java 8 parses as:

    {a=ConfigDelayedMerge(1,${b}), b=ConfigReference(${a})}

Does the ordering of `a` and `b` matter here?

If so, then that's the problem, and if not then the problem must be somewhere in ResolveContext.resolve or a method it calls.